### ISSUE TITLE
Rewrite target paths

### ISSUE THREAD

**Issue by twpayne** [twpayne/chezmoi#2273]
## Is your feature request related to a problem? Please describe.

Some files are in different directories on different machines, for example user-installed fonts are in `~/.local/share/fonts` on Linux and `~/Library/Fonts` on macOS. It would be nice to have a way to handle these differences that is easier to use than [the current method](https://www.chezmoi.io/user-guide/manage-machine-to-machine-differences/#handle-different-file-locations-on-different-systems-with-the-same-contents).

## Describe the solution you'd like

From [this comment](https://github.com/twpayne/chezmoi/issues/2138#issuecomment-1159313813), roughly it could look something like:

1. In your config file, you specify something like:

    ```toml
    [rewriteTargetPaths]
        "Documents" = "One Drive - {organization}/Documents"
    ```

2. This would tell chezmoi that whenever it finds `Documents` in your source state it should replace it with `One Drive - {organization}/Documents` in the target state.

3. To handle machine-to-machine differences, there's the [existing config file generation mechanism](https://www.chezmoi.io/reference/special-files-and-directories/chezmoi-format-tmpl/) which we can use. So, for example, your config file template would look something like:

    ```
    [rewriteTargetPaths]
        "Documents" = "One Drive - {{ registry "HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer\User Shell Folders\Personal" }}"
    ```

   (assuming the addition of a `registry` template function to get values from the Windows registry)

4. For the [use case of different directories in the FAQ entry](https://www.chezmoi.io/user-guide/manage-machine-to-machine-differences/#ignore-files-or-a-directory-on-different-machines), the config file template would look something like:

    ```
    {{ if .chezmoi.os == "darwin" }}
    # config file directory locations are different on macOS (darwin)
    [rewriteTargetPaths]
        ".config/app" = "Library/Application Support/App"
    {{ end }}
    ```

Implementing this has some subtleties (e.g. what should happen if multiple source directories are re-written to a single target directory) and will likely be non-trivial to implement.

## Describe alternatives you've considered

Keeping things as they are.

## Additional context

None.

**Comment by k4lizen** [comment]
I would also really like this feature since programs often have whole folders for configuration files. This person especially seems to be struggling https://stackoverflow.com/a/78245168 xd

**Comment by Ra2-IFV** [comment]
I was wondering if it's possible to have a new attribute for files to handle this, for example `folder_foo.tmpl` with content `bar` in that file, will tell chezmoi to create a folder with contents from `bar` directory.
Might be too confusing for users?

**Comment by Rnbsov** [comment]
No way there is still no solution after 3 years

**Comment by twpayne** [comment]
Are you proposing to implement it? Or are you expecting someone else to do it for you for free?

**Comment by Kotletka89** [comment]
> Are you proposing to implement it? Or are you expecting someone else to do it for you for free?

yeah, why to develop open source at all! let's make this repo private, and integrate subscription system

**Comment by halostatue** [comment]
> > Are you proposing to implement it? Or are you expecting someone else to do it for you for free?
> 
> yeah, why to develop open source at all! let's make this repo private, and integrate subscription system

[Open source maintainers owe you nothing](https://mikemcquaid.com/open-source-maintainers-owe-you-nothing/). Consider this a participation warning. If you cannot participate constructively, you may find yourself blocked from participating at all in this project.

So far, that this is still open suggests that it's a complex issue that would require (a) substantial design to ensure that it's done correctly, (b) an extensive re-think of chezmoi's state management, (c) even more extensive tests to ensure that existing features still work as expected‡, and (d) a migration plan for users if this changes features that are currently heavily used now.

Tom and his collaborators work on chezmoi because we benefit from it and enjoy giving away that work. Tom has _explicitly_ not enabled any sponsorship features and wants no donations for this, because it isn't his full time job. All he asks from participants in this project is a level of respect when interacting with his project and a willingness to be patient for things which are hard.

‡ This is harder than it looks. There are issues that exist based on how two different features interact with one another, and resolving them is non-trivial and sometimes not easily reproducible.

**Comment by twpayne** [comment]
Thank you everyone for contributing here. I'm locking this issue for a few days to give everybody (including me) a bit of time to cool down.

**Comment by Ra2-IFV** [comment]
Sorry for being off-topic, but open source doen't mean asking the developers for doing anything for free. It's "FREEDOM", it allows others help with the project, it hears **reasonable** and **practicial** ideas from users, yet depends on the developers.

The passion and hard work of open source software developers is not free.

**Comment by Lockszmith-GH** [comment]
I've been trying to tackle this myself, and I came up with a somewhat convoluted solution.
It might be a starting point, but it feels way too convoluted right now - functional - yes, maintainable - not really.

It's based on what I call 'bootstrap chezmoi' (read #4415 for more info and link to my dotfiles repo).

I have the different environments in separate dirs in the git repo, my bootstrap configuration sets .chezmoiroot based on whatever logic that makes sense.

## What's relevant to this discussion:
When sharing files across dirs, I use symlinks inside the repo.

## The limitations:
As described earlier by k4lizen, `chezmio edit` doesn't work, and `chezmoi add` overwrites the symlinks with new content - practically breaking the link.

I've created a function that 'takes over' `chezmio edit` and just open the editor after resolving the path with `source-path` - not exactly the same, but it does preserve the symlinks, as it follows them instead of replacing them.

It's also an additional layer to take care of when renaming/moving files.
I wrote a script that resets the symlinks, but it's still a pain.

A possible solution I can think of, is some specific template directive that can allow the code to change the destination path - it will make maintainablility more complex.

The best solution I can think off - is an additional file, a template, that evaluates into path.
Maybe name it <file_name>.pathtmpl

**Comment by byguid** [comment]
One common use case for this feature is to handle different config paths on Windows and Linux (`%appdata%` vs `~/.config`).
For apps like Neovim with a lot of config files, the currently available official solution is not practical.

The simplest and most effective approach for me was to use junctions on Windows that point to `~/.config`.
While not the most robust or universal solution, it works surprisingly well for many everyday use cases.
Best of all, the entire setup can be wrapped into a single script that should be easy to maintain.

```powershell
# run_after__windows-config-junctions.ps1.tmpl @ chezmoi source-path
{{- if eq .chezmoi.os "windows" }}
$paths = @(
    @{ Junction = "$env:LOCALAPPDATA\nvim"; Source = "$HOME\.config\nvim" },
    @{ Junction = "$env:APPDATA\yazi"; Source = "$HOME\.config\yazi" }
    # >> add more paths here (don't forget to append comma on the line above)
)

foreach ($path in $paths) {
    if (-Not (Test-Path $path.Junction)) {
        New-Item -ItemType Junction -Path $path.Junction -Value $path.Source
    }
}
{{- end }}
```

Hope some people find this helpful. 

And a big thank you to the authors of chezmoi ❤️ (beautiful name, by the way, once you hear pronunciation)

**Comment by cspotcode** [comment]
> Implementing this has some subtleties (e.g. what should happen if multiple source directories are re-written to a single target directory) and will likely be non-trivial to implement.

To implement this iteratively, and keep the future design space maximally open, can we raise an error when multiple source directories are re-written to a single target directory? (I'll refer to this as "source-merging," for lack of a better name)

This way, anyone who does not need source-merging is able to use `rewriteTargetPaths`, and we avoid the need to implement any sort of logic to handle this case, other than detecting and raising the error.  This also keeps the door open for implementing source-merging in the future without causing backwards-compatibility headaches.

The intent of this error is to communicate:
"You cannot do this today.  You *might* be able to in the future, but semantics have not been designed, so nobody knows what that would look like."

**Comment by A2va** [comment]
I'm not entirely sure this is related to this issue, but from what I have tested chezmoi does always place the `dot_config` folder under `$HOME/.config` even if `XDG_CONFIG_HOME` was set.

**Comment by halostatue** [comment]
> I'm not entirely sure this is related to this issue, but from what I have tested chezmoi does always place the `dot_config` folder under `$HOME/.config` even if `XDG_CONFIG_HOME` was set.  

This is expected behaviour. `chezmoi apply` is not aware of XDG_CONFIG_HOME and will not retarget .config if it is set.

**Comment by twpayne** [comment]
To add to what halostatue said, when you have a directory called `dot_config` in your source directory, this means exactly the `.config` directory in your target (home) directory.

chezmoi respects the XDG base directory specification for its own configuration and cache.

You're also correct that this issue ("rewrite target paths") is related to moving directories around. For example, when this issue is fixed you should be able to write something like this in your config file template:

```
[rewriteTargetPaths]
    ".config" = {{ env "XDG_CONFIG_HOME" | quote }}
```
